HeterogeneousCore/CUDAUtilities/src/CUDAEventCache.cc

```diff
@@ -1,20 +1,30 @@
 #include "HeterogeneousCore/CUDAUtilities/interface/CUDAEventCache.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/cudaCheck.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/currentDevice.h"
+#include "HeterogeneousCore/CUDAUtilities/interface/ScopedSetDevice.h"
+
+#include <cuda/api_wrappers.h>
 
 namespace cudautils {
+  void CUDAEventCache::Deleter::operator()(cudaEvent_t event) const {
+    if (device_ != -1) {
+      ScopedSetDevice deviceGuard{device_};
+      cudaCheck(cudaEventDestroy(event));
+    }
+  }
+
   // CUDAEventCache should be constructed by the first call to
   // getCUDAEventCache() only if we have CUDA devices present
   CUDAEventCache::CUDAEventCache() : cache_(cuda::device::count()) {}
 
-  std::shared_ptr<cuda::event_t> CUDAEventCache::getCUDAEvent() {
-    return cache_[cuda::device::current::get().id()].makeOrGet([]() {
-      auto current_device = cuda::device::current::get();
-      // We should not return a recorded, but not-yet-occurred event
-      return std::make_unique<cuda::event_t>(
-          current_device
-              .create_event(  // default; we should try to avoid explicit synchronization, so maybe the value doesn't matter much?
-                  cuda::event::sync_by_busy_waiting,
-                  // it should be a bit faster to ignore timings
-                  cuda::event::dont_record_timings));
+  SharedEventPtr CUDAEventCache::getCUDAEvent() {
+    const auto dev = cudautils::currentDevice();
+    return cache_[dev].makeOrGet([dev]() {
+      // TODO(?): We should not return a recorded, but not-yet-occurred event
```

Comment:
do you mean that we should check the state of an event before reusing it ?

Reply:
The comment was there already, I just added a "TODO" as I didn't fully remember if that is in issue or not (and being afraid it is). At the moment all practical uses of the events are "occurred" before they are returned to the cache.

But yes, in general a check should be added.